### DISCUSSION TITLE
Download: Random changes in element ordering

### DISCUSSION THREAD

**Discussion by d41k4n** [Dynatrace/dynatrace-configuration-as-code#304] [Archive] [answered]
It seems the ordering of JSON array elements is not deterministic and may differ between downloads even though the config didn't actually change for the tenant/environment at all. This makes diffing and versioning of downloaded configs difficult if not impossible.

Examples of JSON arrays whose elements appear in a non-deterministic/random order with each subsequent config download;

- `propagationTypes` (e.g. management-zone, auto-tag configs)
- `tags` (e.g. synthetic-monitor, maintenance-window configs)
- `tagFilters` (e.g. anomaly-detection-metrics config)
- `managementZones` (e.g. synthetic-monitor config)
- `pathDefinitions` (calculated-metrics-log config)
- `hostFilters` (calculated-metrics-log config)
- `sourceEntities` (calculated-metrics-log config)

Is this a general/known limitation of the Dynatrace API?

Also, the generated config yaml suffers from a similar issue where the sequence of referenced JSON configs may change without any .json file being added or removed... Again, this makes diffing and version control of downloaded configs a challenge....

I understand the main goal of the tool is the automated config deployment at scale where the download function is currently more of a "goodie" in order to get started with creating templates, but nonetheless I think this could be improved going forward...

**Comment by ghost** [comment, marked as answer]
Hi about the random order for the elements in the json or in the yaml file, this is a behaviour from the current dynatrace api (I tested this with the api endpoint management zones) , this might be cover in the new api v2. @tatomir146  this might be a good input for the api meeting
